core/include/cond.h

```diff
@@ -0,0 +1,223 @@
+ *     while (queue_length(&pipe->queue) == MAX_QUEUE_LENGTH && !pipe->closed) {
+ *         cond_wait(&pipe->write_cond, &pipe->lock);
+ *     }
+ *     if (pipe->closed) {
+ *         mutex_unlock(&pipe->lock);
+ *         return 0;
+ *     }
+ *     add_to_queue(&pipe->queue, c);
+ *     cond_signal(&pipe->read_cond);
+ *     mutex_unlock(&pipe->lock);
+ *     return 1;
+ * }
+ *
+ * void pipe_close(pipe_t* pipe) {
+ *     mutex_lock(&pipe->lock);
+ *     pipe->closed = true;
+ *     cond_broadcast(&pipe->read_cond);
+ *     cond_broadcast(&pipe->write_cond);
+ *     mutex_unlock(&pipe->lock);
+ * }
+ *
+ * int pipe_read(pipe_t* pipe, char* buf) {
+ *     mutex_lock(&pipe->lock);
+ *     while (queue_length(&pipe->queue) == 0 && !pipe->closed) {
+ *         cond_wait(&pipe->read_cond, &pipe->lock);
+ *     }
+ *     if (pipe->closed) {
+ *         mutex_unlock(&pipe->lock);
+ *         return 0;
+ *     }
+ *     *buf = remove_from_queue(&pipe->queue);
+ *     cond_signal(&pipe->write_cond);
+ *     mutex_unlock(&pipe->lock);
+ *     return 1;
+ * }
+ *
+ * Note that this could actually be written with a single condition variable.
+ * However, the example includes two for didactic reasons.
+ *
+ * @author      Sam Kumar <samkumar@berkeley.edu>
+ */
+
+#ifndef COND_H
+#define COND_H
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "list.h"
+#include "mutex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Condition variable structure. Must never be modified by the user.
+ */
+typedef struct {
+    /**
+     * @brief   The process waiting queue of the condition variable.
+     * @internal
+     */
+    list_node_t queue;
+} cond_t;
+
+/**
+ * @brief   Static initializer for cond_t.
+ * @details This initializer is preferable to cond_init().
+ */
+#define COND_INIT { { NULL } }
+
+/**
+ * @brief Initializes a condition variable.
+ * @details For initialization of variables use COND_INIT instead.
+ *          Only use the function call for dynamically allocated condition
+ *          variables.
+ * @param[out] cond    pre-allocated condition structure, must not be NULL.
+ */
+static inline void cond_init(cond_t *cond)
+{
+    cond->queue.next = NULL;
+}
+
+/**
+ * @brief Waits on a condition.
+ *
+ * @param[in] cond          Condition variable to wait on.
+ * @param[in] mutex         Mutex object held by the current thread.
+ */
+void cond_wait(cond_t *cond, mutex_t *mutex);
+
+/**
+ * @brief Wake up threads waiting on a condition variable.
+ * @details For commit purposes you should probably use cond_signal() and
+ *          cond_broadcast() instead.
+ *
+ * @param[in] cond          Condition variable whose threads to wake up.
+ * @param[in] broadcast     If true, wakes up all threads. If false, only
+ *                          wakes up the first waiting thread.
+ */
+void _cond_signal(cond_t *cond, bool broadcast);
+
+
+/**
+ * @brief Wakes up one thread waiting on the condition variable. The thread is
```

Comment:
this isn't very brief anymore, I suggest to add an empty line after the first sentence. So that the longer description will go into the detailed section, same applies below.